File: firmware/remote-trigger/Doorbell.cpp
#include "Doorbell.h"

Doorbell::Doorbell(uint8_t pin, uint16_t bit_length, uint16_t frame_spacing, bool *frame, uint8_t length) {
  _pin = pin;
  _frame = frame;
  _bit_count = length;
  _bit_length = bit_length;
  _frame_spacing = frame_spacing;

  PinMode(_pin, OUTPUT);
}

void Doorbell::ring(uint8_t times) {
  for (uint8_t i = 0; i < times; i++) {
    // Ring button
    sendFrame();
    delayMicroseconds(_frame_spacing);
  }
}

void Doorbell::sendFrame() {
  for (uint8_t i = 0; i < _bit_count; i++) {
    digitalWrite(_pin, _frame[i]);
    delayMicroseconds(_bit_length);
  }
}

File: firmware/micropython/remote-trigger/scripts/lib/doorbell.py
from time import sleep_us
from machine import Pin

class Doorbell:
    _frame = [
      0, 1, 0, 1, 0, 0, 1,
      1, 0, 0, 1, 0, 1, 0,
      1, 0, 1, 1, 0, 1, 0,
      1, 0, 0, 1, 1, 0, 0,
      1, 0, 1, 0, 1, 0, 1
    ]
    # (short_pulse_duration, long_pulse_duration)
    _pulse_timings = (475, 1435)
    _frame_spacing = 8667

    def __init__(self, pin):
        self.pin = Pin(pin, Pin.OUT)


    def send_frame(self):
        for f in self._frame:
            self.pin.toggle()
            sleep_us(self._pulse_timings[f])
        self.pin.value(0)


    def ring(self, times):
        for i in range(times):
            self.send_frame()
            sleep_us(self._frame_spacing)
